core/src/main/java/org/apache/iceberg/view/ViewMetadata.java

```diff
@@ -117,13 +128,172 @@ default ViewMetadata checkAndNormalize() {
+      this.history = Lists.newArrayList(base.history());
+      this.properties = Maps.newHashMap(base.properties());
+      this.changes = Lists.newArrayList(base.changes());
+    }
+
+    public Builder upgradeFormatVersion(int newFormatVersion) {
+      Preconditions.checkArgument(
+          newFormatVersion >= this.formatVersion,
+          "Cannot downgrade v%s view to v%s",
+          formatVersion,
+          newFormatVersion);
+
+      if (this.formatVersion == newFormatVersion) {
+        return this;
+      }
+
+      this.formatVersion = newFormatVersion;
+      this.changes.add(new MetadataUpdate.UpgradeFormatVersion(newFormatVersion));
+      return this;
+    }
+
+    public Builder setLocation(String newLocation) {
+      if (null != this.location && this.location.equals(newLocation)) {
+        return this;
+      }
+
+      this.location = newLocation;
+      this.changes.add(new MetadataUpdate.SetLocation(newLocation));
+      return this;
+    }
+
+    public Builder setCurrentVersionId(int newVersionId) {
+      if (this.currentVersionId == newVersionId) {
+        return this;
+      }
+
+      this.currentVersionId = newVersionId;
+      this.changes.add(new MetadataUpdate.SetCurrentViewVersion(newVersionId));
+      return this;
+    }
+
+    public Builder setCurrentSchemaId(Integer newSchemaId) {
+      if (null != this.currentSchemaId && this.currentSchemaId.equals(newSchemaId)) {
+        return this;
+      }
+
+      this.currentSchemaId = newSchemaId;
+      this.changes.add(new MetadataUpdate.SetCurrentSchema(newSchemaId));
+      return this;
+    }
+
+    public Builder addSchema(Schema schema) {
+      this.schemas.add(schema);
+      this.changes.add(new MetadataUpdate.AddSchema(schema, schema.highestFieldId()));
+      return this;
+    }
+
+    public Builder setSchemas(Iterable<Schema> schemasToAdd) {
+      for (Schema schema : schemasToAdd) {
+        addSchema(schema);
+      }
+
+      return this;
+    }
+
+    public Builder addVersion(ViewVersion version) {
+      this.versions.add(version);
+      this.changes.add(new MetadataUpdate.AddViewVersion(version));
+      this.history.add(
+          ImmutableViewHistoryEntry.builder()
+              .timestampMillis(version.timestampMillis())
+              .versionId(version.versionId())
+              .build());
+      return this;
+    }
+
+    public Builder setProperties(Map<String, String> updated) {
+      if (updated.isEmpty()) {
+        return this;
+      }
+
+      this.properties.putAll(updated);
+      this.changes.add(new MetadataUpdate.SetProperties(updated));
+      return this;
+    }
+
+    public Builder removeProperties(Set<String> propertiesToRemove) {
+      if (propertiesToRemove.isEmpty()) {
+        return this;
+      }
+
+      propertiesToRemove.forEach(this.properties::remove);
+      this.changes.add(new MetadataUpdate.RemoveProperties(propertiesToRemove));
+      return this;
+    }
+
+    public ViewMetadata build() {
```

Comment:
Shouldn't this also enforce the number of versions? That way we don't need the validation that also modifies the object that is constructed?

Reply:
the problem with enforcing the number of versions in the `build()` method is that we'd have to enforce this in some other place as well, because `ViewMetadataParser` doesn't go through the builder, but we want to have the same behavior, independent from whether an instance is created through the builder or directly via `ImmutableViewMetadata.of(...)`